presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java

```diff
@@ -0,0 +1,482 @@
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Optional<AggregateInfo>> context)
+        {
+            // optimize if and only if
+            // some aggregation functions have a distinct mask symbol
+            // and if not all aggregation functions on same distinct mask symbol (this case handled by SingleDistinctOptimizer)
+            Set<Symbol> masks = ImmutableSet.copyOf(node.getMasks().values());
+            if (masks.size() != 1 || node.getMasks().size() == node.getAggregations().size()) {
+                return context.defaultRewrite(node, Optional.empty());
+            }
+
+            AggregateInfo aggregateInfo = new AggregateInfo(node.getGroupBy(),
+                    Iterables.getOnlyElement(masks),
+                    node.getAggregations(),
+                    node.getFunctions());
+
+            PlanNode source = context.rewrite(node.getSource(), Optional.of(aggregateInfo));
+
+            // Change aggregate node to do second aggregation, handles this part of optimized plan mentioned above:
+            //          SELECT a1, a2,..., an, arbitrary(f1 if group = 0 else null),...., arbitrary(fm if group = 0 else null), F(c if group = 1 else null)
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            for (Map.Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                FunctionCall functionCall = entry.getValue();
+                if (entry.getValue().isDistinct()) {
+                    aggregations.put(entry.getKey(), new FunctionCall(functionCall.getName(),
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(Iterables.getOnlyElement(aggregateInfo.getNewDistinctAggregateSymbols().values()).toSymbolReference())));
+                    functions.put(entry.getKey(), node.getFunctions().get(entry.getKey()));
+                }
+                else {
+                    // Aggregations on non-distinct are already done by new node, just extract the non-null value
+                    Symbol argument = aggregateInfo.getNewNonDistinctAggregateSymbols().get(entry.getKey());
+                    QualifiedName functionName = QualifiedName.of("arbitrary");
+                    aggregations.put(entry.getKey(), new FunctionCall(functionName,
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(argument.toSymbolReference())));
+                    functions.put(entry.getKey(),
+                            metadata.getFunctionRegistry().resolveFunction(functionName, ImmutableList.of(symbolAllocator.getTypes().get(argument).getTypeSignature()), false));
```

Comment:
This is a little hard to read. I'd either introduce a variable for the parameter types or the function or format it across multiple lines.